src/Common/SpaceSaving.h

```diff
@@ -228,76 +235,96 @@ class SpaceSaving
-            counter_list.resize(m_capacity);
+            for (size_t i = 0; i < alpha_map.size(); i++)
+                alpha_map[i] = std::min(alpha_map[i] + rhs.alpha_map[i], MAX_ALPHA_VALUE);
         }
 
-        for (size_t i = 0; i < counter_list.size(); ++i)
-            counter_list[i]->slot = i;
-        rebuildCounterMap();
+        truncateIfNeeded(true);
     }
 
     std::vector<Counter> topK(size_t k) const
     {
-        std::vector<Counter> res;
+        std::vector<Counter> new_list;
+        new_list.reserve(counter_list.size());
         for (auto & counter : counter_list)
-        {
-            res.push_back(*counter);
-            if (res.size() == k)
-                break;
-        }
-        return res;
+            new_list.push_back(counter);
+
+        size_t return_size = std::min(new_list.size(), k);
+        ::partial_sort(
+            new_list.begin(), new_list.begin() + return_size, new_list.end(), [](const auto & l, const auto & r) { return l > r; });
+
+        if (k < new_list.size())
+            new_list.resize(k);
+
+        return new_list;
     }
 
     void write(WriteBuffer & wb) const
     {
-        writeVarUInt(size(), wb);
+        std::vector<Counter> new_list;
+        new_list.reserve(counter_list.size());
         for (auto & counter : counter_list)
-            counter->write(wb);
+            new_list.push_back(counter);
+
+        size_t return_size = std::min(new_list.size(), capacity());
+        ::partial_sort(
+            new_list.begin(), new_list.begin() + return_size, new_list.end(), [](const auto & l, const auto & r) { return l > r; });
+
+        writeVarUInt(return_size, wb);
+        for (size_t i = 0; i < return_size; i++)
+            new_list[i].write(wb);
 
         writeVarUInt(alpha_map.size(), wb);
         for (auto alpha : alpha_map)
             writeVarUInt(alpha, wb);
     }
 
-    void read(ReadBuffer & rb)
+    void read(ReadBuffer & rb, size_t capacity)
     {
-        destroyElements();
         size_t count = 0;
         readVarUInt(count, rb);
+        if (count > capacity)
+            throw DB::Exception(
+                DB::ErrorCodes::TOO_LARGE_ARRAY_SIZE, "Found too large when reading counters (Passed: {}. Maximum: {})", count, capacity);
 
+        if (!count)
+        {
+            /// It is possible that the state was written before the set was initialized, which would add 0 Counters and the alpha map
+            /// The size of the alpha map depends on the version and what was the default but we can just ignore it (no Counters mean there
+            /// should be no errors in the alpha map)
+            skipAlphaMap(rb);
+            return;
+        }
+
+        resize(capacity);
         for (size_t i = 0; i < count; ++i)
         {
-            auto counter = std::make_unique<Counter>();
-            counter->read(rb);
-            counter->hash = counter_map.hash(counter->key);
+            auto counter = Counter();
```

Comment:
This object looks like a bunch of int-s and, in practice, isn't movable. I'm not sure how hot this place is, but we could create it via `counter_list.emplace()` and avoid copying it later on.